cpp/src/arrow/compute/kernels/aggregate_test.cc

```diff
@@ -1321,5 +1321,288 @@ TEST_F(TestVarStdKernelIntegerLength, Basics) {
+
+template <typename ArrowType>
+class TestFloatingQuantileKernel : public TestPrimitiveQuantileKernel<ArrowType> {};
+
+template <typename ArrowType>
+class TestInt64QuantileKernel : public TestPrimitiveQuantileKernel<ArrowType> {};
+
+#define INTYPE(x) Datum(static_cast<typename TypeParam::c_type>(x))
+#define DOUBLE(x) Datum(static_cast<double>(x))
+// output type per interplation: linear, lower, higher, nearest, midpoint
+#define O(a, b, c, d, e) \
+  { DOUBLE(a), INTYPE(b), INTYPE(c), INTYPE(d), DOUBLE(e) }
+// output type same as input if only 0 and 1 quantiles are calculated
+#define I(a, b, c, d, e) \
+  { INTYPE(a), INTYPE(b), INTYPE(c), INTYPE(d), INTYPE(e) }
+
+TYPED_TEST_SUITE(TestIntegerQuantileKernel, IntegralArrowTypes);
+TYPED_TEST(TestIntegerQuantileKernel, Basics) {
+  // reference values from numpy
+  // ordered by interpolation method: {linear, lower, higher, nearest, midpoint}
+  this->AssertQuantileIs("[1]", 0.1, O(1, 1, 1, 1, 1));
+  this->AssertQuantileIs("[1, 2]", 0.5, O(1.5, 1, 2, 1, 1.5));
+  this->AssertQuantileIs("[3, 5, 2, 9, 0, 1, 8]", 0.5, O(3, 3, 3, 3, 3));
+  this->AssertQuantileIs("[3, 5, 2, 9, 0, 1, 8]", 0.33, O(1.98, 1, 2, 2, 1.5));
+  this->AssertQuantileIs("[3, 5, 2, 9, 0, 1, 8]", 0.9, O(8.4, 8, 9, 8, 8.5));
+  this->AssertQuantilesAre("[3, 5, 2, 9, 0, 1, 8]", {0.5, 0.9},
+                           {O(3, 3, 3, 3, 3), O(8.4, 8, 9, 8, 8.5)});
+  this->AssertQuantilesAre("[3, 5, 2, 9, 0, 1, 8]", {1, 0.5},
+                           {O(9, 9, 9, 9, 9), O(3, 3, 3, 3, 3)});
+  this->AssertQuantileIs("[3, 5, 2, 9, 0, 1, 8]", 0, I(0, 0, 0, 0, 0));
+  this->AssertQuantileIs("[3, 5, 2, 9, 0, 1, 8]", 1, I(9, 9, 9, 9, 9));
+  this->AssertQuantilesAre("[3, 5, 2, 9, 0, 1, 8]", {1, 0},
+                           {I(9, 9, 9, 9, 9), I(0, 0, 0, 0, 0)});
+  this->AssertQuantilesAre(
+      "[3, 5, 2, 9, 0, 1, 8]", {1, 0, 0, 1},
+      {I(9, 9, 9, 9, 9), I(0, 0, 0, 0, 0), I(0, 0, 0, 0, 0), I(9, 9, 9, 9, 9)});
+
+  this->AssertQuantileIs("[5, null, null, 3, 9, null, 8, 1, 2, 0]", 0.21,
+                         O(1.26, 1, 2, 1, 1.5));
+  this->AssertQuantilesAre("[5, null, null, 3, 9, null, 8, 1, 2, 0]", {0.5, 0.9},
+                           {O(3, 3, 3, 3, 3), O(8.4, 8, 9, 8, 8.5)});
+  this->AssertQuantilesAre("[5, null, null, 3, 9, null, 8, 1, 2, 0]", {0.9, 0.5},
+                           {O(8.4, 8, 9, 8, 8.5), O(3, 3, 3, 3, 3)});
+
+  this->AssertQuantileIs({"[5]", "[null, null]", "[3, 9, null]", "[8, 1, 2, 0]"}, 0.33,
+                         O(1.98, 1, 2, 2, 1.5));
+  this->AssertQuantilesAre({"[5]", "[null, null]", "[3, 9, null]", "[8, 1, 2, 0]"},
+                           {0.21, 1}, {O(1.26, 1, 2, 1, 1.5), O(9, 9, 9, 9, 9)});
+
+  this->AssertQuantilesEmpty("[]", {0.5});
+  this->AssertQuantilesEmpty("[null, null, null]", {0.1, 0.2});
+  this->AssertQuantilesEmpty({"[null, null]", "[]", "[null]"}, {0.3, 0.4});
+}
+
+#ifndef __MINGW32__
```

Comment:
Why was this necessary? Can you point to a specific error?

Reply:
Below are errors from unit test.
First two errors may be caused by index not calculated precisely, pick the wrong side from adjacent data points.
For other failed tests, the printed values are same, looks quantile value is not calculate precisely.

```
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -9
  numeric_scalar->value
    Which is: -inf
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: inf
  numeric_scalar->value
    Which is: 11
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: 7
  numeric_scalar->value
    Which is: 7
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
[  FAILED  ] TestFloatingQuantileKernel/0.Floats, where TypeParam = arrow::FloatType (1 ms)
[----------] 1 test from TestFloatingQuantileKernel/0 (1 ms total)

[----------] 1 test from TestFloatingQuantileKernel/1, where TypeParam = arrow::DoubleType
[ RUN      ] TestFloatingQuantileKernel/1.Floats
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -9
  numeric_scalar->value
    Which is: -inf
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: inf
  numeric_scalar->value
    Which is: 11
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: 7
  numeric_scalar->value
    Which is: 7
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1353: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -3.5
  numeric_scalar->value
    Which is: -3.5
[  FAILED  ] TestFloatingQuantileKernel/1.Floats, where TypeParam = arrow::DoubleType (1 ms)
[----------] 1 test from TestFloatingQuantileKernel/1 (1 ms total)

[----------] 1 test from TestInt64QuantileKernel/0, where TypeParam = arrow::Int64Type
[ RUN      ] TestInt64QuantileKernel/0.Int64
[       OK ] TestInt64QuantileKernel/0.Int64 (0 ms)
[----------] 1 test from TestInt64QuantileKernel/0 (0 ms total)

[----------] 2 tests from TestRandomQuantileKernel
[ RUN      ] TestRandomQuantileKernel.Normal
[       OK ] TestRandomQuantileKernel.Normal (7 ms)
[ RUN      ] TestRandomQuantileKernel.Overlapped
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/aggregate_test.cc:1361: Failure
Expected equality of these values:
  quantiles[j]
    Which is: -47.0185
  numeric_scalar->value
    Which is: -47.0185
[  FAILED  ] TestRandomQuantileKernel.Overlapped (13 ms)
[----------] 2 tests from TestRandomQuantileKernel (20 ms total)

[----------] Global test environment tear-down
[==========] 168 tests from 133 test suites ran. (385 ms total)
[  PASSED  ] 165 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] TestFloatingQuantileKernel/0.Floats, where TypeParam = arrow::FloatType
[  FAILED  ] TestFloatingQuantileKernel/1.Floats, where TypeParam = arrow::DoubleType
[  FAILED  ] TestRandomQuantileKernel.Overlapped
```

Reply:
Looks mingw32 is still using 80-bit x87 fpu for floating point calculation, while other platforms/compilers uses 64-bit floating point. Might be the reason why many floating point tests failed in arrow unit tests?

Some similar issues:
https://www.raspberrypi.org/forums/viewtopic.php?t=205567
https://stackoverflow.com/questions/13447444/what-difference-between-vs-c-and-mingw-to-implement-double-type

Reply:
Yes, I think that is due to 80-bit FP (double rounding?). However, this doesn't explain the infinity results...

Reply:
I didn't investigate deep.
The test array is `[-9, 7, Inf, -Inf, 2, 11]`. So `-inf` and `-9` are two smallest values,  `11` and `inf` are two biggest values.
I guess the error comes from quantile to index conversion, index = quantile * (size - 1). The index is between two data points, -9 and -inf, maybe approaching 0.5, and lead to different result for `nearest interpolation` method.

Reply:
Ok, that sounds plausible. Let's not worry too much about it. 32-bit x86 platforms are really legacy anyway.